src/VisualStudio/Core/Def/Implementation/Remote/ServiceHubRemoteHostClient.cs

```diff
@@ -270,6 +244,52 @@ private void OnGlobalOperationStopped(object sender, GlobalOperationEventArgs e)
             }
         }
 
+        private void RegisterPersistentStorageLocationServiceChanges()
+        {
+            var persistentStorageLocationService = this.Workspace.Services.GetService<IPersistentStorageLocationService>();
+            if (persistentStorageLocationService != null)
+            {
+                persistentStorageLocationService.StorageLocationChanging += OnPersistentStorageLocationServiceStorageLocationChanging;
+
+                EnqueueStorageLocationChange(Workspace.CurrentSolution.Id, persistentStorageLocationService.TryGetStorageLocation(Workspace.CurrentSolution.Id));
+            }
+        }
+
+        private void OnPersistentStorageLocationServiceStorageLocationChanging(object sender, PersistentStorageLocationChangingEventArgs e)
+        {
+            EnqueueStorageLocationChange(e.SolutionId, e.NewStorageLocation);
+
+            if (e.MustUseNewStorageLocationImmediately)
+            {
+                _currentRemoteWorkspaceNotificationTask.Wait();
+            }
+        }
+
+        private void EnqueueStorageLocationChange(SolutionId solutionId, string storageLocation)
+        {
+            lock (_currentRemoteWorkspaceNotificationTaskGate)
+            {
+                _currentRemoteWorkspaceNotificationTask = _currentRemoteWorkspaceNotificationTask.SafeContinueWithFromAsync(_ =>
+                {
+                    return _rpc.InvokeAsync(
+                        nameof(IRemoteHostService.UpdateSolutionStorageLocation),
+                        new object[] { solutionId, storageLocation });
+                }, CancellationToken.None, TaskScheduler.Default);
```

Comment:
Should we support cancellation here so the wait in Unregister can be shorter?

Reply:
We want to ensure that register/unregister things are happening cleanly and in order. The suspicion is the wait, in practice, is effectively zero, and only happens in cases like renaming solutions.